man/fakechroot.1
.\" Automatically generated by Pod::Man 4.10 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "FAKECHROOT 1"
.TH FAKECHROOT 1 "16 Mar 2019" "fakechroot" " "
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
fakechroot \- gives a fake chroot environment
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBfakechroot\fR
[\fB\-s\fR|\fB\-\-use\-system\-libs\fR]
[\fB\-l\fR|\fB\-\-lib\fR\ \fIlibrary\fR]
[\fB\-d\fR|\fB\-\-elfloader\fR\ \fIldso\fR]
[\fB\-e\fR|\fB\-\-environment\fR\ \fItype\fR]
[\fB\-c\fR|\fB\-\-config\-dir\fR\ \fIdirectory\fR]
[\fB\-b\fR|\fB\-\-bindir\fR\ \fIdirectory\fR]
[\fB\-\-\fR]
[\fIcommand\fR]
.PP
\&\fBfakechroot\fR
\&\fB\-h\fR|\fB\-\-help\fR
.PP
\&\fBfakechroot\fR
\&\fB\-v\fR|\fB\-\-version\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
fakechroot runs a command in an environment where is additional possibility to
use \fBchroot\fR\|(8) command without root privileges. This is useful for allowing
users to create own chrooted environment with possibility to install another
packages without need for root privileges.
.PP
fakechroot replaces some C library functions (\fBchroot\fR\|(2), \fBopen\fR\|(2), etc.)
by ones that simulate the effect of being called with root privileges.
.PP
These wrapper functions are provided as a shared library \fIlibfakechroot.so\fR
which is loaded through the \f(CW\*(C`LD_PRELOAD\*(C'\fR mechanism of the dynamic loader.
(See \fBld.so\fR\|(8))
.PP
In fake chroot you can install i.e. Debian bootstrap with \fBdebootstrap\fR\|(8)
command. In such environment you can run i.e. \fBapt\-get\fR\|(8) command to install
another packages. You don't need a special privileges and you can run it on
common user's account.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-b\fR|\fB\-\-bindir\fR \fIdirectory\fR" 4
.IX Item "-b|--bindir directory"
Specify a directory which contains the replacement executables
shipped with fakechroot.
Those are script files with the extension \f(CW\*(C`.fakechroot\*(C'\fR.
By default they are distributed over the
\&\f(CW\*(C`bin\*(C'\fR and \f(CW\*(C`sbin\*(C'\fR directories in fakechroot's install prefix,
but sometimes a fixed location at build time is not feasible.
.IP "\fB\-c\fR|\fB\-\-config\-dir\fR \fIdirectory\fR" 4
.IX Item "-c|--config-dir directory"
Specify a directory which contains additional configuration for fakechroot. The
default directory are \f(CW\*(C`$HOME/.fakechroot\*(C'\fR and \f(CW\*(C`/etc/fakechroot\*(C'\fR.
.IP "\fB\-d\fR \fIldso\fR|\fB\-\-elfloader\fR \fIldso\fR" 4
.IX Item "-d ldso|--elfloader ldso"
Specify an alternative dynamic linker. This dynamic linker will be invoked
directly.
.IP "\fB\-e\fR|\fB\-\-environment\fR \fItype\fR" 4
.IX Item "-e|--environment type"
Load additional configuration with environment. This configuration file
is a shell script which is executed before calling \fIcommand\fR. The
script can set additional environment variables, like i.e.:
\&\f(CW\*(C`FAKECHROOT_EXCLUDE_PATH\*(C'\fR, \f(CW\*(C`FAKECHROOT_CMD_SUBST\*(C'\fR or \f(CW\*(C`LD_LIBRARY_PATH\*(C'\fR.
.Sp
The environment type is guessed based on command name with optional extension
removed (e.g. running \fIgettext.sh\fR loads \f(CW\*(C`gettext\*(C'\fR environment file). If
\&\fIcommand\fR argument is \fBfakeroot\fR\|(1) this argument is omitted and next argument
is taken as environment type.
.Sp
The configuration file name is \fI\fItype\fI.env\fR and is searched at
\&\fI\f(CI$HOME\fI/.fakechroot\fR and \fI/etc/fakechroot\fR directories.
.Sp
The default environment type is \fBdefault\fR and its configuration file name is
\&\f(CW\*(C`default.env\*(C'\fR.
.Sp
The special environment \fBnone\fR means that no environment settings are loaded
at all.
.IP "\fB\-l\fR \fIlibrary\fR|\fB\-\-lib\fR \fIlibrary\fR" 4
.IX Item "-l library|--lib library"
Specify an alternative wrapper library. The default is \fIlibfakechroot.so\fR
.IP "\fB\-h\fR|\fB\-\-help\fR" 4
.IX Item "-h|--help"
Display help.
.IP "\fB\-s\fR|\fB\-\-use\-system\-libs\fR" 4
.IX Item "-s|--use-system-libs"
Use system libraries before chroot's libraries. This might be a workaround if
system dynamic linker cannot load \fIlibc.so\fR from fake chroot.
.Sp
Try this setting if you noticed following errors:
.Sp
.Vb 4
\& $ fakechroot /usr/sbin/chroot /tmp/sarge /bin/true
\& /bin/true: relocation error: /srv/sarge/lib/tls/libc.so.6: symbol _dl
\& _starting_up, version GLIBC_PRIVATE not defined in file ld\-linux.so.2
\&  with link time reference
\&
\& $ fakechroot /usr/sbin/chroot /tmp/centos4 /bin/true
\& Segmentation fault
.Ve
.IP "\fB\-v\fR|\fB\-\-version\fR" 4
.IX Item "-v|--version"
Display version.
.IP "[\fB\-\-\fR] \fIcommand\fR" 4
.IX Item "[--] command"
Any command you want to be run as fakechroot. Use '\fB\-\-\fR' if in the command
you have other options that may confuse fakechroot's option parsing.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
An example session with fakechroot:
.PP
.Vb 2
\&  $ id
\&  uid=1000(dexter) gid=1000(dexter) groups=1000(dexter)
\&
\&  $ fakechroot fakeroot debootstrap sid /tmp/sid
\&  I: Retrieving Release 
\&  I: Retrieving Release.gpg 
\&  I: Checking Release signature
\&  ...
\&  I: Base system installed successfully.
\&
\&  $ fakechroot fakeroot chroot /tmp/sid apt\-get install \-q hello
\&  Reading package lists...
\&  Building dependency tree...
\&  Reading state information...
\&  The following NEW packages will be installed:
\&    hello
\&  0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
\&  Need to get 57.4 kB of archives.
\&  After this operation, 558 kB of additional disk space will be used.
\&  Get:1 http://ftp.us.debian.org/debian/ sid/main hello amd64 2.8\-4 [57.4 kB]
\&  Fetched 57.4 kB in 0s (127 kB/s)
\&  Selecting previously unselected package hello.
\&  (Reading database ... 24594 files and directories currently installed.)
\&  Unpacking hello (from .../archives/hello_2.8\-4_amd64.deb) ...
\&  Processing triggers for man\-db ...
\&  Processing triggers for install\-info ...
\&  Setting up hello (2.8\-4) ...
\&
\&  $ fakechroot chroot /tmp/sid hello
\&  Hello, world!
.Ve
.SH "FAKEROOT"
.IX Header "FAKEROOT"
\&\fBfakeroot\fR\|(1) is a complementary tool which emulates root environment. fakeroot
and fakechroot might wrap the same C library functions, i.e. \fBmknod\fR\|(2)
function. It is important to start fake environment in proper order. fakeroot
should be started inside fakechroot:
.PP
.Vb 1
\&  $ fakechroot fakeroot chroot /tmp/sid /bin/mknod /tmp/device c 1 2
.Ve
.SH "SECURITY ASPECTS"
.IX Header "SECURITY ASPECTS"
fakechroot is a regular, non-setuid program. It does not enhance a user's
privileges.
.PP
fakechroot should not be used as a tool for enhancing system security i.e. by
separating (sandboxing) applications. It is very easy to escape from a fake
chroot environment.
.PP
fakechroot should not be run with real root privileges. It might decrease the
security of the system because the fakechroot provides own version of core
functions with behavior depended on some environment variables.
.SH "FILES"
.IX Header "FILES"
.IP "\fIlibfakechroot.so\fR" 4
.IX Item "libfakechroot.so"
The shared library containing the wrapper functions.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.IP "\fB\s-1FAKECHROOT\s0\fR" 4
.IX Item "FAKECHROOT"
The value is true for fake chroot environment.
.IP "\fB\s-1FAKECHROOT_AF_UNIX_PATH\s0\fR" 4
.IX Item "FAKECHROOT_AF_UNIX_PATH"
The root directory for unix sockets. The default value is the same as
\&\f(CW\*(C`FAKECHROOT_BASE\*(C'\fR and it can be set separately if the \f(CW\*(C`FAKECHROOT_BASE\*(C'\fR is
too long and the unix socket path could exceed the limit of \fB108\fR bytes.
.IP "\fB\s-1FAKECHROOT_BASE\s0\fR" 4
.IX Item "FAKECHROOT_BASE"
The root directory of fake chroot environment.
.IP "\fB\s-1FAKECHROOT_CMD_SUBST\s0\fR" 4
.IX Item "FAKECHROOT_CMD_SUBST"
A list of command substitutions. If a program tries to execute one of
the commands given (path relative to the chroot, trailing dot is removed) then
the substitute command runs instead (path to substitute command is not
chrooted).
.Sp
The substituted command inherits \f(CW\*(C`FAKECHROOT_*\*(C'\fR variables but the original
\&\f(CW\*(C`FAKECHROOT_BASE\*(C'\fR variable which is saved as \f(CW\*(C`FAKECHROOT_BASE_ORIG\*(C'\fR. It
means that substituted command runs outside fakechroot environment. Also
original command name is saved as \f(CW\*(C`FAKECHROOT_CMD_ORIG\*(C'\fR.
.Sp
For example:
.Sp
.Vb 1
\&  export FAKECHROOT_CMD_SUBST=/usr/bin/mkfifo=/bin/true
.Ve
.Sp
will substitute \f(CW\*(C`/bin/true\*(C'\fR for \f(CW\*(C`/usr/bin/mkfifo\*(C'\fR and will make possible to
install sysvinit binary package.
.Sp
Give as many substitute commands as you want, separated by \f(CW\*(C`:\*(C'\fR
(colon) characters.
.Sp
It is suggested to substitute at least:
.RS 4
.IP "\(bu" 2
\&\f(CW\*(C`/bin/mount=/bin/true\*(C'\fR
.IP "\(bu" 2
\&\f(CW\*(C`/sbin/insserv=/bin/true\*(C'\fR
.IP "\(bu" 2
\&\f(CW\*(C`/sbin/ldconfig=/bin/true\*(C'\fR
.IP "\(bu" 2
\&\f(CW\*(C`/usr/bin/env=/usr/bin/env.fakechroot\*(C'\fR
.IP "\(bu" 2
\&\f(CW\*(C`/usr/bin/ischroot=/bin/true\*(C'\fR
.IP "\(bu" 2
\&\f(CW\*(C`/usr/bin/ldd=/usr/bin/ldd.fakechroot\*(C'\fR
.IP "\(bu" 2
\&\f(CW\*(C`/usr/bin/mkfifo=/bin/true\*(C'\fR
.RE
.RS 4
.Sp
to make \fBdebootstrap\fR\|(8) working correctly.
.Sp
To prevent some looping, the command substitution is done only if
\&\f(CW\*(C`FAKECHROOT_CMD_ORIG\*(C'\fR variable is not set currently.
.RE
.IP "\fB\s-1FAKECHROOT_DEBUG\s0\fR" 4
.IX Item "FAKECHROOT_DEBUG"
The fakechroot library will dump some debugging info if this variable is set.
.IP "\fB\s-1FAKECHROOT_DETECT\s0\fR" 4
.IX Item "FAKECHROOT_DETECT"
If this variable is set then \f(CW\*(C`fakechroot \f(CIversion\f(CW\*(C'\fR string is printed to
standard output and the current process is terminated with status taken from
this variable. It can be a method to check if fakechroot is preloaded
correctly.
.Sp
.Vb 1
\&  $ case "\`FAKECHROOT_DETECT=1 /bin/echo\`" in fakechroot*) echo LOADED;; esac
.Ve
.IP "\fB\s-1FAKECHROOT_ELFLOADER\s0\fR" 4
.IX Item "FAKECHROOT_ELFLOADER"
A path to another dynamic linker (i.e. \fI/lib/ld\-linux.so.2\fR for i386
architecture, \fI/lib64/ld\-linux\-x86\-64.so.2\fR for x86_64 architecture).
.Sp
This dynamic linker will be invoked directly. The dynamic linker don't allow
to change \f(CW\*(C`argv[0]\*(C'\fR besides the file name of the executable file, so some
application won't work correctly, i.e. \fBbusybox\fR\|(1).
.IP "\fB\s-1FAKECHROOT_EXCLUDE_PATH\s0\fR" 4
.IX Item "FAKECHROOT_EXCLUDE_PATH"
The list of directories which are excluded from being chrooted. The elements
of list are separated with colon.
.Sp
The \fI/dev\fR, \fI/proc\fR and \fI/sys\fR directories are excluded by default if this
environment variable is not set.
.Sp
This list has to contain at most 100 elements.
.IP "\fB\s-1FAKECHROOT_EXTRA_LIBRARY_PATH\s0\fR" 4
.IX Item "FAKECHROOT_EXTRA_LIBRARY_PATH"
The list of extra directories in fake chroot environment that are added to
\&\f(CW\*(C`LD_LIBRARY_PATH\*(C'\fR variable. The directories might be used by some important
commands which use libraries placed in the \fIrunpath\fR. In that case dynamic
linker can't find correct directory in a fake chroot environment without the
extra library path.
.Sp
The default value is \f(CW\*(C`/lib/systemd:/usr/lib/man\-db\*(C'\fR for \fBsystemctl\fR\|(1) and
\&\fBman\fR\|(1) commands.
.IP "\fB\s-1FAKECHROOT_VERSION\s0\fR" 4
.IX Item "FAKECHROOT_VERSION"
The version number of the current fakechroot library.
.IP "\fB\s-1LD_LIBRARY_PATH\s0\fR, \fB\s-1LD_PRELOAD\s0\fR" 4
.IX Item "LD_LIBRARY_PATH, LD_PRELOAD"
Fakechroot is implemented by wrapping system calls. This is accomplished by
setting \f(CW\*(C`LD_PRELOAD=libfakechroot.so\*(C'\fR. If this library can't be found by
dynamic linker, the \fI/etc/ld.so.conf\fR file or \f(CW\*(C`LD_LIBRARY_PATH\*(C'\fR variable

NEWS
== Version ?.?? ==

?? ??? ????

* The +ldd+(1) wrapper can work if overriden with +FAKECHROOT_CMD_SUBST+
  environment variable.
* The new +env+(1) wrapper was added.  It preserves fakechroot environment even
  for +--ignore-environment+ option.
* The special environment +none+ means that no environment settings are loaded
  at all.
* Fixes were made for +readline+(2) function if destination path is similar to
  +FAKECHROOT_PATH+.
* Fixes were made for +mktemp+(3) function if used on a path in
  +FAKECHROOT_EXCLUDE_PATHS+.
* The +_xftw64+(glibc) function is reenabled. It had wrong wrapper.
* Fixes were made for +__realpath_chk+(glibc) function when +__chk_fail+
  function is missing.
* New functions +mkostemp+(3), +mkostemp64+(3), +mkostemps+(3),
  +mkostemps64+(3), +mkstemps+(3) and +mkstemps64+(3) were implemented.
  It fixes +sed -i+ command.

== Version 2.16 ==

11 Dec 2011

* The fakechroot script loads additional environment settings from
  configuration directory (+--config-dir+ option). By default additional
  settings are provided for +chroot+(8) and +debootstrap+(8) commands.
* Wrapped +chroot+(8) command loads +ld.so.conf+ paths to +LD_LIBRARY_PATH+
  environment variable.
* The +debootstrap+(8) command works with default, buildd and minbase
  variants.
* Fixes were made for +getpeeraddr+(3) and +getsockaddr+(3) functions.
  +lwp-request+ command is working correctly.

== Version 2.15 ==

29 Sep 2011

* New function +faccessat+(2) was added.  It fixes +test -r+ command.
  Thanks to Johannes Schauer.
* The +popen+(3) function were reimplemented based on OpenBSD source to
  prevent some coredumps with newer GNU C Library.

== Version 2.14 ==

18 Dec 2010

* The source code was refactored: all functions was moved to separated files.
* The +opendir+(3) function is compiled only if it doesn't call other function
  internally.  It fixes +opendir+(3), +fts_open+(3) and +ftw+(3) functions.
* The +fts_*+(3) functions were reimplemented based on OpenBSD source.
* The +ftw_*+(3) functions were reimplemented based on GNU C Library source.
* The +__opendir2+(3) function was reimplemented based on FreeBSD source.
* Fixes were made for older GNU C Library and for cross-compiling environment.

== Version 2.13.1 ==

30 Nov 2010

* Fixes were made for older GNU C Library.

== Version 2.13 ==

29 Nov 2010

* Fixes were made for +realpath+(3) and +ftw+(3) functions.  +apt-ftparchive+
  command currently is working.
* Fixes were made for +canonicalize_file_name+(3), +ftw64+(3), +nftw+(3) and
  +nftw64+(3) functions.
* New functions +fts_children+(3) and +fts_read+(3) were added.
* Lazy dynamic symbol loading was implemented.  Thanks to Alexander Shishkin.
* Fixed bug when +FAKECHROOT_EXCLUDE_PATH+ list was sometimes ignored.
* All source code is now on LGPL license again.

== Version 2.12 ==

10 Nov 2010

* New function +execlp+(3).  Thanks to Robert Spanton.
* New functions +statfs+(2) and +statvfs+(2).  Thanks to Paweł Stawicki.
* Better support for GNU/kFreeBSD was made.

== Version 2.11 ==

12 Sep 2010

* Fixes were made for +getpeername+(2) and +getsockname+(2) functions.
  +host 127.0.0.1+ command currently is working.  Thanks to Daniel Tschan.
* Fixes were made for +canonicalize_file_name+(3) function.  +man-db+ command
  currently is working.
* New functions +__*_chk+(3) were added for GNU libc with fortify +level > 0+.
* New environment variable +FAKECHROOT_AF_UNIX_PATH+ defines an optional
  prefix for unix sockets.
* Better support for FreeBSD was made.
* Autoconf scripts were refactored.
* More test units were added.

== Version 2.10 ==

25 Aug 2010

* Handle properly GNU libc +open+(2), +open64+(2), +openat+(2) and
  +openat64+(2) inline functions.
* Compatibility were improved for +scandir+(3) and +scandir64+(3) functions
  on the latest GNU libc.
* Fixed +lstat+(2) function returns correct +st_size+.  Thanks to Daniel
  Baumann.
* Fixed +readlink+(2) function.  Correctly truncates content if buffer is too
  small.
* Fixed +undefined symbol: rpl_malloc+ error for compiled 32-bit library on
  64-bit architecture by dropping +AC_FUNC_MALLOC+ in +configure.ac+.
* Fixed +chroot+(2) function for path which contains dot character.  Thanks to
  Robie Basak.
* New function +utimensat+(2) fixes +cp+ command on Debian sid.
* New functions +linkat+(2), +mknodat+(2), +mkfifoat+(2), +readlinkat+(2)
  and +symlinkat+(2).
* New functions +popen+(3) and +system+(3) for GNU libc.
* +dpkg-shlibdeps(1)+ works again because +cd $FAKECHROOT_BASE+ does not
  change current working directory to root directory.
* New +FAKECHROOT_CMD_SUBST+ environment variable handles a list of
  substituted commands.  Thanks to Richard W.M. Jones.
* New scripts ldd and ldconfig with .fakechroot suffix are installed.
* New +make test+ target uses Perl's +prove+ tool to run test units.

== Version 2.9 ==

31 Mar 2009

* Fixes were made for +getpeername+(2) and +getsockname+(2) functions.  Thanks to
  Axel Thimm and Fedora people.
* Fixes were made for +execve+(2):
  * Always copy necessary variables to new environment.
  * Do not make duplicates of environment variables.  Thanks to Richard
    W.M. Jones.
* Fixed were made for +chroot+(2).  It is possible to escape chroot.  Thanks
  to Richard W.M. Jones.
* Fixed were made for +mktemp+(2).  There was a buffer overflow.  Thanks to
  Mikhail Gusarov.
* New function: +futimesat+(2).  Fixes +touch -m+ command.
* New functions: +bindtextdomain+(3), +inotify_add_watch+(2).
* More test units was added, +make check+ works as expected.

== Version 2.8 ==

25 Jul 2008

* Fixes were made for +__fxstatat64+(3) function which broke +chown+ command
  on i386 architecture.
* Better support for FreeBSD was made.

== Version 2.7.1 ==

17 Jul 2008

* Compiles with older GNU libc: readlink type of return value is detected.
* Compiles with uClibc: existence of getwd function is detected.

== Version 2.7 ==

15 Jul 2008

* Stability and compatibility were improved for the latest libc.

scripts/fakechroot.sh
#!@SHELL@

# fakechroot
#
# Script which sets fake chroot environment
#
# (c) 2011, 2013 Piotr Roszatycki <dexter@debian.org>, LGPL


FAKECHROOT_VERSION=@VERSION@


fakechroot_die () {
    echo "$@" 1>&2
    exit 1
}


fakechroot_usage () {
    fakechroot_die "Usage:
    fakechroot [-l|--lib fakechrootlib]
               [-d|--elfloader ldso]
               [-s|--use-system-libs]
               [-e|--environment type]
               [-c|--config-dir directory]
               [-b|--bindir directory]
               [--] [command]
    fakechroot -v|--version
    fakechroot -h|--help"
}


fakechroot_next_cmd () {
    if [ "$1" = "fakeroot" ]; then
        shift
        # skip the options
        while [ $# -gt 0 ]; do
            case "$1" in
                -h|-v)
                    break
                    ;;
                -u|--unknown-is-real)
                    shift
                    ;;
                -l|--lib|--faked|-s|-i|-b)
                    shift 2
                    ;;
                --)
                    shift
                    break
                    ;;
                *)
                    break
                    ;;
            esac
        done
    fi

    if [ -n "$1" -a "$1" != "-v" -a "$1" != "-h" ]; then
        fakechroot_environment=`basename -- "$1"`
    fi
}


if [ "$FAKECHROOT" = "true" ]; then
    fakechroot_die "fakechroot: nested operation is not supported"
fi


# fakechroot doesn't work with CDPATH correctly
unset CDPATH


# Default settings
fakechroot_lib=libfakechroot.so
fakechroot_paths=@libpath@
fakechroot_sysconfdir=@sysconfdir@
fakechroot_confdir=
fakechroot_environment=
fakechroot_bindir=

if [ "$fakechroot_paths" = "no" ]; then
    fakechroot_paths=
fi

if command -v which >/dev/null; then
    fakechroot_echo=`which echo`
    fakechroot_echo=${fakechroot_echo:-@ECHO@}
else
    fakechroot_echo=@ECHO@
fi


# Get options
fakechroot_getopttest=`getopt --version`
case $fakechroot_getopttest in
    getopt*)
        # GNU getopt
        fakechroot_opts=`getopt -q -l lib: -l elfloader: -l use-system-libs -l config-dir: -l environment: -l bindir: -l version -l help -- +l:d:sc:e:b:vh "$@"`
        ;;
    *)
        # POSIX getopt ?
        fakechroot_opts=`getopt l:d:sc:e:b:vh "$@"`
        ;;
esac

if [ "$?" -ne 0 ]; then
    fakechroot_usage
fi

eval set -- "$fakechroot_opts"

while [ $# -gt 0 ]; do
    fakechroot_opt=$1
    shift
    case "$fakechroot_opt" in
        -h|--help)
            fakechroot_usage
            ;;
        -v|--version)
            echo "fakechroot version $FAKECHROOT_VERSION"
            exit 0
            ;;
        -l|--lib)
            fakechroot_lib=`eval echo "$1"`
            fakechroot_paths=
            shift
            ;;
        -d|--elfloader)
            FAKECHROOT_ELFLOADER=$1
            export FAKECHROOT_ELFLOADER
            shift
            ;;
        -s|--use-system-libs)
            fakechroot_paths="${fakechroot_paths:+$fakechroot_paths:}/usr/lib:/lib"
            ;;
        -c|--config-dir)
            fakechroot_confdir=$1
            shift
            ;;
        -e|--environment)
            fakechroot_environment=$1
            shift
            ;;
        -b|--bindir)
            fakechroot_bindir=$1
            shift
            ;;
        --)
            break
            ;;
    esac
done

if [ -z "$fakechroot_environment" ]; then
    fakechroot_next_cmd "$@"
fi


# Autodetect if dynamic linker supports --argv0 option
if [ -n "$FAKECHROOT_ELFLOADER" ]; then
    fakechroot_detect=`$FAKECHROOT_ELFLOADER --argv0 echo $fakechroot_echo yes 2>&1`
    if [ "$fakechroot_detect" = yes ]; then
        FAKECHROOT_ELFLOADER_OPT_ARGV0="--argv0"
        export FAKECHROOT_ELFLOADER_OPT_ARGV0
    fi
fi


# Swap libfakechroot and libfakeroot in LD_PRELOAD if needed
# libfakeroot must come first
# an alternate fakeroot library may be given
# in the FAKEROOT_ALT_LIB environment variable
if [ -n "$FAKEROOT_ALT_LIB" ]; then
    lib_libfakeroot="$FAKEROOT_ALT_LIB"
else
    lib_libfakeroot="libfakeroot-sysv.so"
fi

for preload in $(echo "$LD_PRELOAD" | tr ':' ' '); do
    case "$preload" in
        "$lib_libfakeroot")
            lib_libfakeroot_to_preload="$preload"
            ;;
        *)
            lib_to_preload="${lib_to_preload:+${lib_to_preload}:}$preload"
            ;;
    esac
done


# Make sure the preload is available
fakechroot_paths="$fakechroot_paths${LD_LIBRARY_PATH:+${fakechroot_paths:+:}$LD_LIBRARY_PATH}"
fakechroot_lib="${lib_libfakeroot_to_preload:+${lib_libfakeroot_to_preload}:}$fakechroot_lib${lib_to_preload:+:$lib_to_preload}"

fakechroot_detect=`LD_LIBRARY_PATH="$fakechroot_paths" LD_PRELOAD="$fakechroot_lib" FAKECHROOT_DETECT=1 $fakechroot_echo 2>&1`
case "$fakechroot_detect" in
    fakechroot*)
        fakechroot_libfound=yes
        ;;
    *)
        fakechroot_libfound=no
esac

if [ $fakechroot_libfound = no ]; then
    fakechroot_die "fakechroot: preload library not found, aborting."
fi


# Additional environment setting from configuration file
if [ "$fakechroot_environment" != "none" ]; then
    for fakechroot_e in "$fakechroot_environment" "${fakechroot_environment%.*}" default; do
        for fakechroot_d in "$fakechroot_confdir" "$HOME/.fakechroot" "$fakechroot_sysconfdir"; do
            fakechroot_f="$fakechroot_d/$fakechroot_e.env"
            if [ -f "$fakechroot_f" ]; then
                . "$fakechroot_f"
                break 2
            fi
        done
    done
fi


# Check if substituted command is called
fakechroot_cmd=`command -v "$1"`

fakechroot_cmd_wrapper=`
    IFS=:
    for fakechroot_cmd_subst in $FAKECHROOT_CMD_SUBST; do
        case "$fakechroot_cmd_subst" in
            "$fakechroot_cmd="*)
                echo "${fakechroot_cmd_subst#*=}"
                break 2
                ;;
        esac
    done
`

# Set FAKECHROOT_CMD_ORIG if wrapped
if [ -n "$fakechroot_cmd_wrapper" ]; then
    FAKECHROOT_CMD_ORIG="$fakechroot_cmd"
    export FAKECHROOT_CMD_ORIG
fi

fakechroot_cmd=${fakechroot_cmd_wrapper:-$1}


# Execute command
if [ -z "$*" ]; then
    LD_LIBRARY_PATH="$fakechroot_paths" LD_PRELOAD="$fakechroot_lib" ${SHELL:-/bin/sh}
    exit $?
else
    if [ -n "$fakechroot_cmd" ]; then
        # Call substituted command
        shift
        LD_LIBRARY_PATH="$fakechroot_paths" LD_PRELOAD="$fakechroot_lib" "$fakechroot_cmd" "$@"
        exit $?
    else
        # Call original command
        LD_LIBRARY_PATH="$fakechroot_paths" LD_PRELOAD="$fakechroot_lib" "$@"
        exit $?
    fi
fi
